sdk/core/System.ClientModel/src/Pipeline/ResponseBufferingPolicy.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// Pipeline policy to buffer response content or add a timeout to response content
+/// managed by the client.
+/// </summary>
+public class ResponseBufferingPolicy : PipelinePolicy
+{
+    internal static readonly ResponseBufferingPolicy Shared = new();
+
+    public ResponseBufferingPolicy()
+    {
+    }
+
+    public sealed override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+        => ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    public sealed override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+        => await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        CancellationToken oldToken = message.CancellationToken;
+        using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(oldToken);
+
+        TimeSpan invocationNetworkTimeout = (TimeSpan)message.NetworkTimeout!;
+        cts.CancelAfter(invocationNetworkTimeout);
+        try
+        {
+            message.CancellationToken = cts.Token;
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline, currentIndex);
+            }
+        }
+        catch (OperationCanceledException ex)
+        {
+            ThrowIfCancellationRequestedOrTimeout(oldToken, cts.Token, ex, invocationNetworkTimeout);
+            throw;
+        }
+        finally
+        {
+            message.CancellationToken = oldToken;
+            cts.CancelAfter(Timeout.Infinite);
+        }
+
+        if (!message.BufferResponse)
+        {
+            return;
```

Comment:
Do we not want to respect network timeout via [ReadTimeoutStream](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/Internal/ReadTimeoutStream.cs) when not buffering?

Reply:
Ah, good catch.  I had intended to move that logic into this PR (and had done so in a separate investigative PR, but forgot here).  Thanks - will bring it in now!